File: LeetCode/java/1054_Distant_Barcodes.java
import java.util.*;

class Solution {
    public int[] rearrangeBarcodes(int[] barcodes) {
        int[] answer = new int[barcodes.length];
        Map<Integer, Integer> map = new HashMap<>();

        for (int n : barcodes) {
            map.put(n, map.getOrDefault(n, 0) + 1);
        }

        List<int[]> list = new ArrayList<>();

        for (int key : map.keySet()) {
            list.add(new int[]{map.get(key), key});
        }

        Collections.sort(list, (x, y) -> {
            if (x[0] < y[0]) {
                return 1;
            }

            if (x[0] > y[0]) {
                return -1;
            }

            return 0;
        });

        int idx = 0;

        for (int i = 0; i < list.size(); i++) {
            int[] n = list.get(i);

            for (int j = 0; j < n[0]; j++) {
                answer[idx] = n[1];

                if (idx == answer.length - 2 || idx == answer.length - 1) {
                    idx = 1;
                } else {
                    idx += 2;
                }
            }
        }

        return answer;
    }
}

File: LeetCode/cpp/1054_Distant_Barcodes.cpp
class Solution {
public:
	vector<int> rearrangeBarcodes(vector<int>& barcodes) {
		vector<int> answer(barcodes.size(), 0);
		unordered_map<int, int> um;

		for (int n : barcodes) {
			um[n]++;
		}

		vector<pair<int, int>> v;

		for (auto m : um) {
			v.push_back({ m.second, m.first });
		}

		sort(v.begin(), v.end(), greater<>());

		int idx = 0;

		for (int i = 0; i < v.size(); i++) {
			for (int j = 0; j < v[i].first; j++) {
				answer[idx] = v[i].second;

				if (idx == answer.size() - 2 || idx == answer.size() - 1) {
					idx = 1;
				}
				else {
					idx += 2;
				}
			}
		}

		return answer;
	}
};